pkg/sql/plpgsql/parser/parse.go

```diff
@@ -126,8 +132,62 @@ func (p *Parser) parse(
 	}, nil
 }
 
+// PlpgSQLStmtCounter is used to accurately report telemetry for plpgsql
+// statements. We can not use the counters due to counters needing to
+// be reset after every statement using reporter.ReportDiagnostics.
+type PlpgSQLStmtCounter map[string]int
+
+func (p *PlpgSQLStmtCounter) String() string {
+	var buf strings.Builder
+	for k, v := range *p {
+		buf.WriteString(fmt.Sprintf("%s: %d\n", k, v))
+
+	}
+	return buf.String()
+}
+
+func ParseAndIncPlpgCounter(sql string, isTest bool) (PlpgSQLStmtCounter, error) {
+	stmtCnt := PlpgSQLStmtCounter{}
+	stmt, err := Parse(sql)
+	if err != nil {
+		return nil, err
+	}
+	for _, s := range stmt.AST.Body {
```

Comment:
I don't think this quite works because statement blocks can be nested. Do you want to give it a shot writing a visitor pattern for it so that we'd be able to walk through the whole AST. For example: https://github.com/cockroachdb/cockroach/blob/ac03a55b391a3c1c1a34d26452e77ed1df18a18a/pkg/sql/sem/tree/walk.go#L1975